pyro/mesh/patch.py

```diff
@@ -186,6 +189,133 @@ def __eq__(self, other):
+
+    Define:
+    r = x
+    theta = y
+    """
+
+    def __init__(self, nx, ny, ng=1,
+                 xmin=0.2, xmax=1.0, ymin=0.0, ymax=1.0):
+
+        # Make sure theta is within [0, PI]
+        assert ymin >= 0.0 and ymax <= np.pi, "y or \u03b8 should be within [0, \u03c0]."
+
+        # Make sure the ghost cells doesn't extend out negative x(r)
+        assert xmin - ng*(xmax-xmin)/nx >= 0.0, \
+            "xmin (r-direction), must be large enough so ghost cell doesn't have negative x."
+
+        super().__init__(nx, ny, ng, xmin, xmax, ymin, ymax)
+
+        self.coord_type = 1
+
+        # Length of the side along r-direction, dr
+
+        self.Lx = ArrayIndexer(np.full((self.qx, self.qy), self.dx),
+                               grid=self)
+
+        # Length of the side along theta-direction, r*dtheta
+
+        self.Ly = ArrayIndexer(np.full((self.qx, self.qy), self.x2d*self.dy),
+                               grid=self)
+
+        # Returns an array of the face area that points in the r(x) direction.
+        # dL_theta x dL_phi = r^2 * sin(theta) * dtheta * dphi
+
+        # dAr_l = - r{i-1/2}^2 * 2pi * cos(theta{i+1/2}) - cos(theta{i-1/2})
+        area_x_l = np.abs(-2.0 * np.pi * self.xl2d.v(buf=self.ng)**2 *
+                 (np.cos(self.yr2d.v(buf=self.ng)) -
+                  np.cos(self.yl2d.v(buf=self.ng))))
+        self.Ax_l = ArrayIndexer(area_x_l, grid=self)
+
+        # dAr_r = - r{i+1/2}^2 * 2pi * cos(theta{i+1/2}) - cos(theta{i-1/2})
+        area_x_r = np.abs(-2.0 * np.pi * self.xr2d.v(buf=self.ng)**2 *
+                 (np.cos(self.yr2d.v(buf=self.ng)) -
+                  np.cos(self.yl2d.v(buf=self.ng))))
+        self.Ax_r = ArrayIndexer(area_x_r, grid=self)
+
+        # Returns an array of the face area that points in the theta(y) direction.
+        # dL_phi x dL_r = dr * r * sin(theta) * dphi
+
+        # dAtheta_l = 0.5 * pi * sin(theta{i-1/2}) * (r{i+1/2}^2 - r{i-1/2}^2)
+        area_y_l = np.abs(0.5 * np.pi * np.sin(self.yl2d.v(buf=self.ng)) *
+                 (self.xr2d.v(buf=self.ng)**2 -
+                  self.xl2d.v(buf=self.ng)**2))
+        self.Ay_l = ArrayIndexer(area_y_l, grid=self)
+
+        # dAtheta_r = 0.5 * pi * sin(theta{i+1/2}) * (r{i+1/2}^2 - r{i-1/2}^2)
+        area_y_r = np.abs(0.5 * np.pi * np.sin(self.yr2d.v(buf=self.ng)) *
+                 (self.xr2d.v(buf=self.ng)**2 -
+                  self.xl2d.v(buf=self.ng)**2))
+        self.Ay_r = ArrayIndexer(area_y_r, grid=self)
+
+        # Returns an array of the volume of each cell.
+        # dV = dL_r * dL_theta * dL_phi
+        #    = (dr) * (r * dtheta) * (r * sin(theta) * dphi)
+        # dV = - 2*np.pi / 3 * (cos(theta{i+1/2}) - cos(theta{i-1/2})) * (r{i+1/2}^3 - r{i-1/2}^3)
+
+        volume = np.abs(-2.0 * np.pi / 3.0 *
+                 (np.cos(self.yr2d.v(buf=self.ng)) - np.cos(self.yl2d.v(buf=self.ng))) *
+                 (self.xr2d.v(buf=self.ng)**3 - self.xl2d.v(buf=self.ng)**3))
```

Comment:
probably doesn't matter for pyro, but in general, we should compute

$$r_r^3 - r_l^3$$

as

$$(r_r - r_l)(r_r^2 + 2 r_l r_r + r_l^2)$$

to minimize roundoff

Reply:
done.